pywr/domains/river.py

```diff
@@ -95,9 +135,138 @@ def __init__(self, model, *args, **kwargs):
         else:
             # reinstate the given cost parameter to pass to the parent constructors
             kwargs["cost"] = cost
+
+        bathymetry = kwargs.pop('bathymetry', None)
+        volume = kwargs.pop('volume', None)
+        level = kwargs.pop('level', None)
+        area = kwargs.pop('area', None)
+        self.weather_cost = kwargs.pop('weather_cost', -999)
+        self.evaporation_cost = kwargs.pop('evaporation_cost', -999)
+        self.rainfall_cost = kwargs.pop('rainfall_cost', -999)
+        const = kwargs.pop('const', 1e6 * 1e-3 * 1e-6)
+
+        # Pywr Storage does not expect a 'weather' kwargs, so move this to instance
+        self.weather = kwargs.pop("weather", None)
+
+
         super(Reservoir, self).__init__(model, *args, **kwargs)
 
 
+        self.const = ConstantParameter(model, const)
+
+        self.rainfall_node = None
+        self.rainfall_recorder = None
+        self.evaporation_node = None
+        self.evaporation_recorder = None
+
+    @classmethod
+    def pre_load(cls, model, data):
+
+        bathymetry = data.pop("bathymetry", None)
+        name = data.pop("name")
+        node = cls(name=name, model=model, **data)
+
+        if bathymetry is not None:
+            if isinstance(bathymetry, str):
+                bathymetry = load_parameter(model, bathymetry)
+                volumes = bathymetry.dataframe['volume'].astype(np.float64)
+                levels = bathymetry.dataframe['level'].astype(np.float64)
+                areas = bathymetry.dataframe['area'].astype(np.float64)
+            else:
+                bathymetry = pd.DataFrame.from_dict(bathymetry)
+                volumes = bathymetry['volume'].astype(np.float64)
+                levels = bathymetry['level'].astype(np.float64)
+                areas = bathymetry['area'].astype(np.float64)
+
+        if volumes is not None and levels is not None:
+            node.level = InterpolatedVolumeParameter(model, node, volumes, levels)
+
+        if volumes is not None and areas is not None:
+            node.area = InterpolatedVolumeParameter(model, node, volumes, areas)
+        if node.weather is not None:
+            node._make_weather_nodes(model, node.weather, node.weather_cost)
+        setattr(node, "_Loadable__parameters_to_load", {})
+        return node
+
+
+    def _make_weather_nodes(self, model, weather, cost):
+
+        if not isinstance(self.area, Parameter):
+            raise ValueError('Weather nodes can only be created if an area Parameter is given.')
+
+        weather = pd.DataFrame.from_dict(weather)
+
+        rainfall = weather['rainfall'].astype(np.float64)
+        evaporation = weather['evaporation'].astype(np.float64)
+
+        self._make_evaporation_node(model, evaporation, cost)
+        self._make_rainfall_node(model, rainfall, cost)
+
+
+    def _make_evaporation_node(self, model, evaporation, cost):
```

Comment:
Rather than a weather attribute, we should have one Rainfall [unit=distance] and one Evaporation [unit=distance] parameter and they should be able to take any Pywr parameter (they will be sometimes constant, monthlyparam, weeklypararm, etc.. or even custom param).

Then as we have, evaporation is done by creating an output node and its max_flow becomes aggregate param which is the Evaporation [unit=distance] param multiplied by the Area param.

For rainfall, we should have a catchment node one by creating an output node and its max_flow becomes aggregate param which is the Rainfall [unit=distance] param multiplied by the Area param.